File: src/main/java/com/github/iamniklas/liocore/led/colorspace/ColorRGB.java
package com.github.iamniklas.liocore.led.colorspace;

import com.github.iamniklas.liocore.led.ColorChannel;
import com.github.iamniklas.liocore.led.typeinterfaces.IConvertibleColor;
import com.github.iamniklas.liocore.led.typeinterfaces.IRGBTypesModifier;

import java.awt.*;

public class ColorRGB implements IRGBTypesModifier<ColorRGB>, IConvertibleColor {
    public static final ColorRGB black =     new ColorRGB(0,     0,  0);
    public static final ColorRGB white =     new ColorRGB(255,   255,255);
    public static final ColorRGB white50 =   new ColorRGB(128,   128,128);
    public static final ColorRGB red =       new ColorRGB(255,   0,  0);
    public static final ColorRGB green =     new ColorRGB(0,     255,0);
    public static final ColorRGB blue =      new ColorRGB(0,     0,  255);
    public static final ColorRGB orange =    new ColorRGB(255,   255,0);
    public static final ColorRGB magenta =   new ColorRGB(255,   0,  255);
    public static final ColorRGB turquoise = new ColorRGB(0,     255,255);

    public static ColorRGB fromSystemColor(LIOColor _systemColor) {
        return new ColorRGB(_systemColor.r, _systemColor.g, _systemColor.b);
    }

    public int r = 255;
    public int g = 255;
    public int b = 255;

    public ColorRGB(int _r, int _g, int _b) {
        r = _r;
        g = _g;
        b = _b;
    }

    public ColorRGB(ColorRGB _color) {
        r = _color.r;
        g = _color.g;
        b = _color.b;
    }

    public ColorRGB filter(ColorChannel _channel) {
        switch (_channel) {
            case Red: return new ColorRGB(0, g, b);
            case Green: return new ColorRGB(r, 0, b);
            case Blue: return new ColorRGB(r, g, 0);
            case Alpha: return new ColorRGB(0, 0, 0);
            default: return this;
        }
    }

    public ColorRGB overrideChannel(ColorChannel _channel, int _value) {
        switch (_channel) {
            case Red: return new ColorRGB(_value, g, b);
            case Green: return new ColorRGB(r, _value, b);
            case Blue: return new ColorRGB(r, g, _value);
            default: return this;
        }
    }

    public ColorRGB cutHigh(ColorChannel _channel, int _value) {
        switch (_channel) {
            case Red: return new ColorRGB(cutHigh(r, _value), g, b);
            case Green: return new ColorRGB(r, cutHigh(g, _value), b);
            case Blue: return new ColorRGB(r, g, cutHigh(b, _value));
            default: return this;
        }
    }
    public ColorRGB cutLow(ColorChannel _channel, int _value) {
        switch (_channel) {
            case Red: return new ColorRGB(cutLow(r, _value), g, b);
            case Green: return new ColorRGB(r, cutLow(g, _value), b);
            case Blue: return new ColorRGB(r, g, cutLow(b, _value));
            default: return this;
        }
    }

    private int cutHigh(int _value, int _cutter) { return Math.min(_value, _cutter); }
    private int cutLow(int _value, int _cutter) {
        return Math.max(_value, _cutter);
    }

    public ColorRGB dim(float _percentage) {
        return new ColorRGB((int)(r * _percentage), (int)(g * _percentage), (int)(b * _percentage));
    }
    public ColorRGB dim(float _r, float _g, float _b) {
        return new ColorRGB((int)(r * _r), (int)(g * _g), (int)(b * _b));
    }

    @Override
    public ColorRGB toRGB() { return this; }
    @Override
    public ColorRGBA toRGBA() { return new ColorRGBA(r, g, b, 255); }
    @Override
    public ColorHSV toHSV() {
        float relativeR = r / 255.0f;
        float relativeG = g / 255.0f;
        float relativeB = b / 255.0f;

        float cMax = Math.max(relativeR, Math.max(relativeG, relativeB));
        float cMin = Math.min(relativeR, Math.min(relativeG, relativeB));

        float cDelta = cMax - cMin;

        //HSV: H
        float h = 0.0f;
        if(cMax == relativeR) {
            h = 60 * (((relativeG - relativeB) / cDelta) % 6);
        } else if(cMax == relativeG) {
            h = 60 * (((relativeB - relativeR) / cDelta) + 2);
        } else if(cMax == relativeB) {
            h = 60 * (((relativeR - relativeG) / cDelta) + 4);
        }

        if(h < 0.0f) {
            h = 360.0f - Math.abs(h);
        }


        //HSV: S
        float s = 0.0f;
        if(cMax != 0.0f) {
            s = cDelta / cMax;
        }


        //HSV: V
        float v = cMax;


        return new ColorHSV((int)h, s, v);
    }
    @Override
    public LIOColor toSystemColor() {
        return new LIOColor(r, g, b);
    }

    @Override
    public String toString() {
        return "ColorRGB{" + "r=" + r + ", g=" + g + ", b=" + b + '}';
    }
}

File: src/main/java/com/github/iamniklas/liocorekotlin/led/ColorRGB.kt
package com.github.iamniklas.liocorekotlin.led

import com.google.gson.annotations.SerializedName
import java.awt.Color

class ColorRGB {
    @SerializedName("R")
    var r = 255

    @SerializedName("G")
    var g = 255

    @SerializedName("B")
    var b = 255

    constructor(_r: Int, _g: Int, _b: Int) {
        r = _r
        g = _g
        b = _b
    }

    constructor(_color: com.github.iamniklas.liocorekotlin.led.ColorRGB) {
        r = _color.r
        g = _color.g
        b = _color.b
    }

    fun filter(_channel: com.github.iamniklas.liocorekotlin.led.ColorChannel?): com.github.iamniklas.liocorekotlin.led.ColorRGB {
        return when (_channel) {
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Red -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                0,
                g,
                b
            )
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Green -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                r,
                0,
                b
            )
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Blue -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                r,
                g,
                0
            )
            else -> com.github.iamniklas.liocorekotlin.led.ColorRGB(0, 0, 0)
        }
    }

    fun setChannel(_channel: com.github.iamniklas.liocorekotlin.led.ColorChannel?, _value: Int) {
        when (_channel) {
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Red -> r = _value
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Green -> g = _value
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Blue -> b = _value
            else -> {
            }
        }
    }

    //Cuts a channel from the lower side - if the channel value is higher than the filter it will be set to 0
    fun cutLow(_channel: com.github.iamniklas.liocorekotlin.led.ColorChannel?, _filter: Int): com.github.iamniklas.liocorekotlin.led.ColorRGB {
        var _filter = _filter
        _filter = Math.max(Math.min(_filter, 255), 0)
        return when (_channel) {
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Red -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                cutLow(r, _filter),
                g,
                b
            )
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Green -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                r,
                cutLow(g, _filter),
                b
            )
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Blue -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                r,
                g,
                cutLow(b, _filter)
            )
            else -> this
        }
    }

    //Cuts a channel from the higher side - if the channel value is lower than the filter it will be set to 0
    fun cutHigh(_channel: com.github.iamniklas.liocorekotlin.led.ColorChannel?, _filter: Int): com.github.iamniklas.liocorekotlin.led.ColorRGB {
        var _filter = _filter
        _filter = Math.max(Math.min(_filter, 255), 0)
        return when (_channel) {
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Red -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                cutHigh(r, _filter),
                g,
                b
            )
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Green -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                r,
                cutHigh(g, _filter),
                b
            )
            com.github.iamniklas.liocorekotlin.led.ColorChannel.Blue -> com.github.iamniklas.liocorekotlin.led.ColorRGB(
                r,
                g,
                cutHigh(b, _filter)
            )
            else -> this
        }
    }

    fun toRGBA(_alpha: Int): com.github.iamniklas.liocorekotlin.led.ColorRGBA {
        return com.github.iamniklas.liocorekotlin.led.ColorRGBA(r, g, b, _alpha)
    }

    fun toHSV(): com.github.iamniklas.liocorekotlin.led.ColorHSV {
        return com.github.iamniklas.liocorekotlin.led.ColorHSV(0, 0.0f, 0.0f)
    }

    private fun cutLow(_value: Int, _filter: Int): Int {
        return if (_value >= _filter) {
            0
        } else {
            _value
        }
    }

    private fun cutHigh(_value: Int, _filter: Int): Int {
        return if (_value <= _filter) {
            0
        } else {
            _value
        }
    }

    fun dim(_percentage: Float): com.github.iamniklas.liocorekotlin.led.ColorRGB {
        return com.github.iamniklas.liocorekotlin.led.ColorRGB(
            (r * _percentage).toInt(),
            (g * _percentage).toInt(),
            (b * _percentage).toInt()
        )
    }

    fun dim(_percentageR: Float, _percentageG: Float, _percentageB: Float): com.github.iamniklas.liocorekotlin.led.ColorRGB {
        return com.github.iamniklas.liocorekotlin.led.ColorRGB(
            (r * _percentageR).toInt(),
            (g * _percentageG).toInt(),
            (b * _percentageB).toInt()
        )
    }

    fun toSystemColor(): Color {
        return Color(r, g, b)
    }

    override fun toString(): String {
        return "$r $g $b"
    }

    companion object {
        val black = com.github.iamniklas.liocorekotlin.led.ColorRGB(0, 0, 0)
        val white = com.github.iamniklas.liocorekotlin.led.ColorRGB(255, 255, 255)
        val white50 = com.github.iamniklas.liocorekotlin.led.ColorRGB(128, 128, 128)
        val red = com.github.iamniklas.liocorekotlin.led.ColorRGB(255, 0, 0)
        val green = com.github.iamniklas.liocorekotlin.led.ColorRGB(0, 255, 0)
        val blue = com.github.iamniklas.liocorekotlin.led.ColorRGB(0, 0, 255)
        val orange = com.github.iamniklas.liocorekotlin.led.ColorRGB(255, 255, 0)
        val magenta = com.github.iamniklas.liocorekotlin.led.ColorRGB(255, 0, 255)
        val torquoise = com.github.iamniklas.liocorekotlin.led.ColorRGB(0, 255, 255)

        fun fromSystemColor(_systemColor: Color): com.github.iamniklas.liocorekotlin.led.ColorRGB {
            return com.github.iamniklas.liocorekotlin.led.ColorRGB(
                _systemColor.red,
                _systemColor.green,
                _systemColor.blue
            )
        }
    }
}